Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,383 @@
+
+            // This is the task that listens to signals
+            group.addTask {
+                for await signal in unixSignals {
+                    return .signalCaught(signal)
+                }
+
+                return .signalSequenceFinished
+            }
+
+            // This is the task that listens to manual graceful shutdown
+            group.addTask {
+                for await _ in gracefulShutdownStream {
+                    return .gracefulShutdownCaught
+                }
+
+                return .gracefulShutdownFinished
+            }
+
+            // This is an optional task that listens to graceful shutdowns from the parent task
+            if let _ = TaskLocals.gracefulShutdownManager {
+                group.addTask {
+                    for await _ in AsyncGracefulShutdownSequence() {
+                        return .gracefulShutdownCaught
+                    }
+
+                    return .gracefulShutdownFinished
+                }
+            }
+
+            // We have to create a graceful shutdown manager per service
+            // since we want to signal them individually and wait for a single service
+            // to finish before moving to the next one
+            var gracefulShutdownManagers = [GracefulShutdownManager]()
+            gracefulShutdownManagers.reserveCapacity(self.services.count)
+
+            for (index, service) in self.services.enumerated() {
+                self.logger.debug(
+                    "Starting service",
+                    metadata: [
+                        self.configuration.logging.serviceKey: "\(service)",
+                    ]
+                )
+
+                let gracefulShutdownManager = GracefulShutdownManager()
+                gracefulShutdownManagers.append(gracefulShutdownManager)
+
+                // This must be addTask and not addTaskUnlessCancelled
+                // because we must run all the services for the below logic to work.
+                group.addTask {
+                    return await TaskLocals.$gracefulShutdownManager.withValue(gracefulShutdownManager) {
+                        do {
+                            try await service.run()
+                            return .serviceFinished(service: service, index: index)
+                        } catch {
+                            return .serviceThrew(service: service, index: index, error: error)
+                        }
+                    }
+                }
+            }
+
+            precondition(gracefulShutdownManagers.count == self.services.count, "We did not create a graceful shutdown manager per service")
+
+            // We are going to wait for any of the services to finish or
+            // the signal sequence to throw an error.
+            while !group.isEmpty {
+                let result: ChildTaskResult? = await group.next()
+
+                switch result {
+                case .serviceFinished(let service, _):
```

Comment:
this may be out of scope, but there could be type of bootstrap tasks that are long running enough so that we would want to cancel them on signal but will finish "early" or iow be short lived

one example i can think of are database migrations, it may take a few minutes and we want to be able to cancel them in an orderly manner, but they have a well defined logical end that is not the entire life of the application. further, we want them to run first, before any other tasks / services run

Reply:
I also still see some use for short running tasks however I couldn't find a concrete example yet. The migration example is IMO not something one would model with a service because arguably the migrations should be run before any of the other services are starting. That's why I would say migrations are run before starting the service runner e.g.

```
try await migrator.run()

let serviceFoo = Foo()
let serviceBar = Bar()

try await serviceRunner([serviceFoo, serviceBar])
```

The current design leaves space for short running service in the future by letting the user pass a service configuration where they can state if a service is long or short running. This is purely additive and I would like to add that once we see a concrete need.